kotlin/authenticate-with-oauth/src/main/java/com/esri/arcgisruntime/sample/authenticatewithoauth/MainActivity.kt

```diff
@@ -0,0 +1,281 @@
+
+  companion object {
+    private val TAG = MainActivity::class.java.simpleName
+  }
+
+  // define configuration for OAuth Portal using custom redirect URL to receive code after auth has been granted
+  private val oAuthConfig: OAuthConfiguration by lazy {
+    OAuthConfiguration(
+      getString(R.string.portal_url),
+      getString(R.string.oauth_client_id),
+      "${BuildConfig.APPLICATION_ID}://${getString(R.string.oauth_redirect_host)}"
+    )
+  }
+
+  // instances of Portal and PortalItem to define what is displayed in the map
+  private val portal: Portal by lazy { Portal(getString(R.string.portal_url)) }
+
+  private val portalItem: PortalItem by lazy { PortalItem(portal, getString(R.string.webmap_world_traffic_id)) }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_main)
+
+    // setup AuthenticationManager to handle auth challenges
+    AuthenticationManager.setAuthenticationChallengeHandler(this)
+    AuthenticationManager.addOAuthConfiguration(oAuthConfig)
+  }
+
+  override fun onResume() {
+    super.onResume()
+    mapView.resume()
+
+    handleIntent(intent)
+
+    mapView.map = ArcGISMap(portalItem)
+  }
+
+  override fun onPause() {
+    mapView.pause()
+    super.onPause()
+  }
+
+  override fun onDestroy() {
+    mapView.dispose()
+    super.onDestroy()
+  }
+
+  /**
+   * Attempt to handle the Intent received by the Activity
+   * If the Intent contains an authorization code, store this in the SharedPreferences
+   *
+   * @param intent the Intent to handle
+   */
+  private fun handleIntent(intent: Intent?) {
+    intent?.authCode?.let { code ->
+      sharedPreferences.putAuthCode(code)
+    }
+  }
+
+  /**
+   * Function for handling authentication challenges.
+   *
+   * If an authorization code exists in the SharedPreferences, it's likely that the user is partially through the OAuth flow
+   * and we need to try to obtain the token by performing a OAuthTokenCredentialRequest.
+   *
+   * If there is not an authorization code in the SharedPreferences, it's likely that this is the user's first attempt
+   * at OAuth. So begin OAuth flow.
+   *
+   * @param authenticationChallenge the authentication challenge to handle
+   * @return the AuthenticationChallengeResponse indicating which action to take
+   */
+  override fun handleChallenge(authenticationChallenge: AuthenticationChallenge?): AuthenticationChallengeResponse {
+    authenticationChallenge?.let { authChallenge ->
+
+      try {
+        // if SharedPreferences has an auth code, we've likely just been through the OAuth flow and now have an auth code
+        // we can use to request a new access token
+        sharedPreferences.authCode?.let {
+          // use the authorization code to get a new access token by executing an OAuthTokenCredentialRequest
+          val request = OAuthTokenCredentialRequest(
```

Comment:
This should be happening in WebView. shouldOverrideUrlLoading()

Reply:
Did we not agree to not use a `WebView` as 1st choice?

Reply:
Yes, it is not. But the whole purpose of doing this is to show a user how they can do it if they want to do it. If the user decides to use the WebBrowser for the challenge then we should be telling them to use the defaultChallengeHandler. 
Besides, the sample currently relies on loading Portal twice so you might not see my point here.

Reply:
https://devtopia.esri.com/runtime/java/issues/426#issuecomment-1703139 referring back to the decisions we made